### DISCUSSION TITLE
How to pass state to form actions

### DISCUSSION THREAD

**Discussion by eakl** [sveltejs/kit#10506] [Q&A] [answered]
## Use case
1. On app load, fetch the user profile that contains the current active workspace of the user
2. On a form action, retrieve the user input and add the current active workspace of the user before adding the record to the database

### Scenario 1
Fetch the user profile in `+hooks.server.ts`, store it in `event.locals` and retrieve the user profile in `routes/post/create/+page.server.ts` form action.
**Problem: The hook runs on each request.**

```ts
// +hooks.server.ts
export const handle: Handle = async ({ event, resolve }) => {
   const user = await fetchUserProfile()
   const postApi = createPostApi()

   event.locals = {
      user,
      postApi
   }
}

// routes/post/create/+page.server.ts
export const actions = {
   default: async ({ request, locals: { user, postApi } }) => {
      const data = await request.formData()
      const { workspace_id } = user

      await postApi.create({
         ...data,
         workspace_id
      })
   }
}
```

### Scenario 2
Fetch the user profile in `+layout.server.ts` and pass it to `routes/+layout.svelte`, set a workspaceId context and retrieve the context in `routes/post/create/+page.svelte`
**Problem: The data can't be passed to form action.**

```ts
// +layout.server.ts
export const load = async () => {
   const user = await fetchUserProfile()

   return {
      user
   }
}

// routes/+layout.svelte
<script lang="ts">
   import { setContext } from 'svelte'
   import { writable } from 'svelte/store'
   import type { LayoutData } from './$types'

   const workspaceId = writable('')

   export let data: LayoutData

   $: workspaceId.set(data.user.workspace_id)

   setContext('workspaceId', workspaceId)
</script>

// routes/post/create/+page.svelte
<script lang="ts">
   import { getContext } from 'svelte'

   const workspaceId = getContext('workspace')
</script>
```

Given that shared stores can't be used with SSR, how can I fetch important user data when the app load for the first time, keep this information in the app state and update the database in a form action with the user context.

**Comment by paoloricciuti** [comment, marked as answer]
Just pass the data to the action with a form and
```html
<input type="hidden" />
```
